_posts/2020-03-04-eda-architecture2.md

```diff
@@ -0,0 +1,73 @@
+
+## Commands, Results and Events
+
+In the same way we have two different kinds of services, we have two different kinds of messages: *Commands* and *Events*.
+
+Events are business messages published on the bus by business services and telling the world what happened.
+
+A business service owns the type of Events it emits. It knows nothing about the services who will process them. It subscribes to the types of Events it needs but knows nothing about their origins.
+
+An Event type defines the meaning of the events of that type and their data schema. They must be documented.
+
+The type of an actual event message is given by its name (a.k.a. *routing key* because it is used by the subscription routing). The event type name must be in the form `SubjectPastParticiple`. For example, `ConversationStarted`, `CustomerCreated`, `ShoppingCartValidated`… If you're not able to immediately give a name to your event type, it means it is not well defined, or that it is not an event. Maybe, you need to refine your service or split it, as you may not have analyzed your value chain deeply enough?
+
+Commands are utility messages consumed by Data Processing Services. Imagine an order you pass to a provider. You don't know who exactly will completed it, you don't know how and when either, but you'll get what you want in you letter box sometime later.
+
+A DP Service owns the types of the commands it consumes and their results. A command is always addressed to the logical service that owns it.
+
+A Command type defines the meaning of the command, its data schema and its result data schema. It must be documented.
+
+The type of an actual command message is given by its name. It is in the form `VerbObject`. For example: `AnnotateText`, `TranscribeAudio`…
+As commands are addressed to a particular logical service, the *routing key* of a command is in the form `logical_service_name.commandname`. For example: `voxo.TranscribeAudio`.
+
+The command contains the return "address" to which the result is to be sent and a reference set by the sender that is returned as-is, along with the command outcome. That reference is called the *correlation identifier* and it is very important for the sender: as all communications are asynchronous, the service requesting the command needs a way to reconcile the received result with the initial request it made.
+
+A Result contains the result of the process, that can be the successful outcome, or an error, and the correlation identifier.
+
+Error results are expected and documented: they are “normal” errors, not bug reports. Bug exceptions *must not return an error result*. In case of unexpected error, the service will replace the input command on the queue once, and if a second try raises an unexpected error again, the message is refused and goes into the dead letter queue for investigation. The exceptions are always logged.
```

Comment:
`the service will replace the input command on the queue once` did you meant `will retry the input command`?
If you really meant `replace`, I'm not sure I understand what you mean.